cpp/src/NVStrings_array.cu

```diff
@@ -20,87 +20,94 @@
+    rmm::device_vector<size_t> sizes(count,0);
+    size_t* d_sizes = sizes.data().get();
+    thrust::for_each_n(execpol->on(0), thrust::make_counting_iterator<unsigned int>(0), count,
+        [d_strings, d_sizes] __device__(unsigned int idx){
+            custring_view* dstr = d_strings[idx];
+            if( dstr )
+                d_sizes[idx] = ALIGN_SIZE(dstr->alloc_size());
+        });
+    // create output object
+    char* d_buffer = pImpl->createMemoryFor(d_sizes);
+    if( d_buffer==0 )
+        return; // this is valid
+    // create offsets
+    rmm::device_vector<size_t> offsets(count,0);
+    thrust::exclusive_scan(execpol->on(0),sizes.begin(),sizes.end(),offsets.begin());
+    // finally, copy the strings
+    custring_view_array d_results = pImpl->getStringsPtr();
+    size_t* d_offsets = offsets.data().get();
+    thrust::for_each_n(execpol->on(0), thrust::make_counting_iterator<unsigned int>(0), count,
+        [d_strings, d_buffer, d_offsets, d_results] __device__(unsigned int idx){
+            custring_view* dstr = d_strings[idx];
+            if( !dstr )
+                return;
+            char* buffer = d_buffer + d_offsets[idx];
+            d_results[idx] = custring_view::create_from(buffer,*dstr);
+        });
+    //
+}
 
 // create a new instance containing only the strings at the specified positions
 // position values can be in any order and can even be repeated
-NVStrings* NVStrings::gather( const int* pos, unsigned int elems, bool bdevmem )
+NVStrings* NVStrings::gather( const int* pos, unsigned int elements, bool bdevmem )
 {
     unsigned int count = size();
-    if( count==0 || elems==0 || pos==0 )
+    if( count==0 || elements==0 || pos==0 )
         return new NVStrings(0);
 
     auto execpol = rmm::exec_policy(0);
     const int* d_pos = pos;
     if( !bdevmem )
     {   // copy indexes to device memory
-        d_pos = const_cast<const int*>(device_alloc<int>(elems,0));
-        CUDA_TRY(cudaMemcpyAsync((void*)d_pos,pos,elems*sizeof(int),cudaMemcpyHostToDevice))
+        d_pos = const_cast<const int*>(device_alloc<int>(elements,0));
+        CUDA_TRY(cudaMemcpyAsync((void*)d_pos,pos,elements*sizeof(int),cudaMemcpyHostToDevice))
     }
-    // get individual sizes
-    rmm::device_vector<long> sizes(elems,0);
-    long* d_sizes = sizes.data().get();
-    custring_view** d_strings = pImpl->getStringsPtr();
-    thrust::for_each_n(execpol->on(0), thrust::make_counting_iterator<unsigned int>(0), elems,
-        [d_strings, d_pos, count, d_sizes] __device__(unsigned int idx){
+    // create working memory
+    rmm::device_vector<custring_view*> results(elements,nullptr);
+    auto d_results = results.data().get();
+    rmm::device_vector<bool> flags(elements,false);
+    auto d_flags = flags.data().get();
+    custring_view_array d_strings = pImpl->getStringsPtr();
+    // do the gather
+    thrust::for_each_n(execpol->on(0), thrust::make_counting_iterator<unsigned int>(0), elements,
```

Comment:
It seems like this should be able to be replaced with a simple `thrust::scatter`.

Reply:
Thanks. Not really part of this PR. It got caught up in some refactoring as part of this PR. Going to leave this one alone for now.